### PR TITLE
New SummaryNode implementation

### PR DESCRIPTION
**Pull request by wito**
This feature implements

1. `Opm::EclIO::SummaryNode`
2. Related changes to `Opm::out::Summary`

`Opm::EclIO::SummaryNode` is a lighter and more focused value object style `struct` than `Opm::SummaryConfigNode`, suitable for use in the Summary reader and writer arenas.

### CODE REVIEW

**Comment by wito** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

File: src/opm/output/eclipse/Summary.cpp [hunk 1]
```diff
@@ -57,76 +57,70 @@
...
 
 namespace {
-    struct ParamCTorArgs
-    {
-        std::string kw;
-        Opm::SummaryConfigNode::Type type;
-    };
-
-    std::vector<ParamCTorArgs> requiredRestartVectors()
+    std::vector<std::pair<std::string, Opm::EclIO::SummaryNode::Type>> requiredRestartVectors()
```
Comment (bska) [inline src/opm/output/eclipse/Summary.cpp]: No.  We're not using `std::pair` for this.  It is infinitely better to have a structure with proper names than to refer to `.first` and `.second`.

File: src/opm/output/eclipse/Summary.cpp [hunk 2]
```diff
@@ -57,76 +57,70 @@
...
+            { "GITH", Type::Total },
         };
     }
 
     std::vector<Opm::SummaryConfigNode>
     requiredRestartVectors(const ::Opm::Schedule& sched)
     {
-        auto entities = std::vector<Opm::SummaryConfigNode>{};
+        std::vector<Opm::SummaryConfigNode> entities {} ;
+        const auto& restartVectors { requiredRestartVectors() } ;
```
Comment (bska) [inline src/opm/output/eclipse/Summary.cpp]: No need to make this a reference.  Just use a regular object.

File: src/opm/output/eclipse/Summary.cpp [hunk 3]
```diff
@@ -1162,41 +1153,29 @@ inline std::vector<Opm::Well> find_wells( const Opm::Schedule& schedule,
         }
 
         return wells;
+
     }
 
-    return {};
+    default:
```
Comment (bska) [inline src/opm/output/eclipse/Summary.cpp]: Preferably not.  Default clauses for switches on enums are more trouble than they're worth.

File: src/opm/output/eclipse/Summary.cpp [hunk 4]
```diff
@@ -1257,16 +1236,17 @@ void eval_udq(const Opm::Schedule& schedule, std::size_t sim_step, Opm::SummaryS
...
 
-    else if (node.category() == Opm::SummaryConfigNode::Category::Group)
-        st.update_group_var(node.namedEntity(), node.keyword(), value);
+    case Opm::EclIO::SummaryNode::Category::Group:
+        return st.update_group_var(node.wgname, node.keyword, value);
 
-    else
-        st.update(node.uniqueNodeKey(), value);
+    default:
+        return st.update(node.unique_key(), value);
```
Comment (bska) [inline src/opm/output/eclipse/Summary.cpp]: `SummaryState::update()` does not return a value.  Pretending that it does will only lead to confusion.  Why do you need to use a switch here anyway.  The original was more readable.

File: src/opm/output/eclipse/Summary.cpp (hunk 1 above)
Comment (wito) [reply src/opm/output/eclipse/Summary.cpp]: Reverted.

File: src/opm/output/eclipse/Summary.cpp (hunk 2 above)
Comment (wito) [reply src/opm/output/eclipse/Summary.cpp]: Corrected.

File: src/opm/output/eclipse/Summary.cpp (hunk 4 above)
Comment (wito) [reply src/opm/output/eclipse/Summary.cpp]: Reverted from `switch`.

File: src/opm/output/eclipse/Summary.cpp (hunk 3 above)
Comment (wito) [reply src/opm/output/eclipse/Summary.cpp]: I'm not sure what you mean by this?

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by wito** [comment]
@joakim-hove That's not really any point right now; there's a test error that I'm in the process of fixing.

File: src/opm/output/eclipse/Summary.cpp (hunk 3 above)
Comment (bska) [reply src/opm/output/eclipse/Summary.cpp]: >> Default clauses for switches on enums are more trouble than they're worth.
>
> I'm not sure what you mean by this?

By adding a `default` clause to a switch on a strong enum you remove the compiler's ability to tell you if there's a case you've forgotten.  This isn't so much an issue when you're writing the code in the first place, but does become a potential error during subsequent maintenance.  I prefer having compile-time warnings/errors to having to do expensive runtime debugging.

**Comment by wito** [comment]
jenkins build this please

File: src/opm/output/eclipse/Summary.cpp (hunk 3 above)
Comment (wito) [reply src/opm/output/eclipse/Summary.cpp]: Addressed.

**Comment by wito** [comment]
jenkins build this please

**Comment by wito** [comment]
@joakim-hove I'm having an issue getting this to build in Jenkins; something about the build system, I think, but I'm not entirely sure what I could have done to upset it. 🙁

**Comment by wito** [comment]
jenkins build this please

**Comment by wito** [comment]
@joakim-hove Looking into it, it seems as this will resolve itself once #1610 is resolved and I can rebase to the new head.

**Comment by bska** [comment]
> it seems as this will resolve itself once #1610 is resolved and I can rebase to the new head.

You need to rebase in any case.  PR #1592 introduced a header and other changes in `Summary.cpp`.

**Comment by wito** [comment]
> > it seems as this will resolve itself once #1610 is resolved and I can rebase to the new head.
> 
> You need to rebase in any case. PR #1592 introduced a header and other changes in `Summary.cpp`.

Indeed I do, but as I'm working on macOS I can't really do so until #1610 is resolved.

**Comment by akva2** [comment]
please confirm https://github.com/OPM/opm-common/pull/1616 takes care of it

**Comment by wito** [comment]
jenkins build this please

**Comment by wito** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
@bska: You had some - reasonable - comments to this PR; could you take a new look and see if it can be merged now?

File: opm/io/eclipse/SummaryNode.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        Count,
+        Mode,
+        Undefined,
+    };
+
+    const std::string keyword;
+    const Category category;
+    const Type type;
+    const std::string wgname;
+    const int number;
```
Comment (bska) [inline opm/io/eclipse/SummaryNode.hpp]: Note that constant datamembers generally disable the move operations and turns a move into a copy.  That's not a problem for arithmetic types, but more worrying for the `string` objects.  Of course, these days most C++ implementations have a small string optimization which usually covers strings of up to 15 characters.  In other words the typical strings used for ECLIPSE keywords and names would fit into the small strings but the principle still applies.

If you really wish to enforce read-only semantics, then my best suggestion is to make `SummaryNode` into a class and then only providing `const` accessor functions.

File: src/opm/io/eclipse/SummaryNode.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        "SURFVISC",
+        "SURFWNUM",
+    } ;
+
+    static const std::regex user_defined_regex { "[ABCFGRSW]U[A-Z]+" } ;
+
+    const bool matched     { std::regex_match(keyword, user_defined_regex) } ;
+    const bool blacklisted { udq_blacklist.find(keyword) != udq_blacklist.end() } ;
+
+    return matched && !blacklisted;
```
Comment (bska) [inline src/opm/io/eclipse/SummaryNode.cpp]: Did you consider writing this as
```C++
return !blacklisted && std::regex_match(keyword, user_defined_regex);
```
?  That way, you don't incur the cost of regular expression matching for blacklisted keywords.

File: src/opm/output/eclipse/Summary.cpp [hunk 7]
```diff
@@ -134,105 +136,86 @@ namespace {
...
+        static const std::vector<ParamCTorArgs> extra_group_vectors {
+            { "GMCTG", Opm::EclIO::SummaryNode::Type::Mode },
+            { "GMCTP", Opm::EclIO::SummaryNode::Type::Mode },
+            { "GMCTW", Opm::EclIO::SummaryNode::Type::Mode },
+        };
+        static const std::vector<ParamCTorArgs> extra_field_vectors {
+            { "FMCTG", Opm::EclIO::SummaryNode::Type::Mode },
+            { "FMCTP", Opm::EclIO::SummaryNode::Type::Mode },
+            { "FMCTW", Opm::EclIO::SummaryNode::Type::Mode },
+        };
```
Comment (bska) [inline src/opm/output/eclipse/Summary.cpp]: Please don't mark these objects as `static`.  The `requiredRestartVectors()` function is called exactly once&mdash;as part of the single-threaded [SMSPEC configuration](https://github.com/OPM/opm-common/blob/d40ebe647a9b808c115daa05e52535004b732919/src/opm/output/eclipse/Summary.cpp#L2478)&mdash;and there's no need to keep the objects in memory for the duration of the process just for that configuration.

File: src/opm/output/eclipse/Summary.cpp [hunk 8]
```diff
@@ -134,105 +136,86 @@ namespace {
...
-
-        auto swfr = SN{ "SWFR", SN::Category::Segment, ::Opm::Location() }
-            .parameterType(SN::Type::Rate);
-
-        auto spr = SN{ "SPR", SN::Category::Segment, ::Opm::Location() }
-            .parameterType(SN::Type::Pressure);
+        std::vector<Opm::EclIO::SummaryNode> ret {};
+
+        static const Opm::EclIO::SummaryNode::Category category { Opm::EclIO::SummaryNode::Category::Segment };
+        static const std::vector<std::pair<std::string,Opm::EclIO::SummaryNode::Type>> requiredVectors {
```
Comment (bska) [inline src/opm/output/eclipse/Summary.cpp]: Please remove `static` here.  The `requiredSegmentVectors()` function is only called during the initial configuration step.

**Comment by bska** [comment]
One more thing.  I didn't expect to see `SummaryNode.[hc]pp` being added to `PUBLIC_*_FILES` twice.  I guess it covers the case
```cmake
ENABLE_ECL_INPUT AND (NOT ENABLE_ECL_OUTPUT)
```
but it still was surprising to me.  Any suggestion on that aspect, @akva2?

**Comment by joakim-hove** [comment]
> This is much better. I just have a couple of more comments and then I think this will be ready for merging.

Thank you for getting back to this - we will fix the remaining points you raise.

**Comment by joakim-hove** [comment]
> I didn't expect to see SummaryNode.[hc]pp being added to PUBLIC_*_FILES twice. 

I was also surprised by that; I concluded it was correct but would be happy to be proven wrong. Time to revisit: https://github.com/OPM/opm-common/pull/1129?

File: src/opm/io/eclipse/SummaryNode.cpp (hunk 6 above)
Comment (wito) [reply src/opm/io/eclipse/SummaryNode.cpp]: I'm not sure, in this case, if that's worth it, given the domain of the function and the relative simplicity of the regular expression.

**Comment by bska** [comment]
jenkins build this please

**Pull request merged**